runtime/src/iree/task/worker.c

```diff
@@ -168,7 +168,18 @@ iree_task_t* iree_task_worker_try_steal_task(iree_task_worker_t* worker,
   if (task) return task;
 
   // If we still didn't steal any tasks then let's try the slist instead.
+  //
+  // Locking worker->local_task_queue.mutex looks out of place here as this is
+  // atomic list pop, but the problem is that we may be racing here against
+  // another thread calling iree_task_queue_flush_from_lifo_slist, copying
+  // linked list elements out of the atomic list, into a plain linked list,
+  // then doing a plain non-atomic pop on that. Fortunately,
+  // iree_task_queue_flush_from_lifo_slist locks worker->local_task_queue.mutex
+  // around that non-atomic operation, so we can avoid the race by also locking
+  // it here.
+  iree_slim_mutex_lock(&worker->local_task_queue.mutex);
```

Comment:
I'm not sure I understand this - the whole point of an atomic slist is that it doesn't need locks. There should not be any possibility of races.

Reply:
(not saying there aren't with the current implementation, but that every atomic slist I've used doesn't need locks like this and that would imply we're doing something else wrong)

Reply:
1. Notice that this file, `worker.c`, is not part of the implementation of the atomic slist, it's a callsite.
2. What's specific about that callsite (as alluded to in the comment being added here) is that it operates on this list here (`iree_atomic_task_slist_pop`) concurrently with another callsite (also in `worker.c` but called on another worker thread) calling `iree_task_queue_flush_from_lifo_slist`.
3. What's specific about `iree_task_queue_flush_from_lifo_slist` is that it performs a shallow-copy of list elements from the atomic slist to a plain list (`iree_atomic_task_slist_flush`) then performs plain accesses on those (`iree_task_list_pop_front`).
4. The `iree_atomic_task_slist_pop` from step 2. above races on the `->next` list element pointer with the `iree_task_list_pop_front` from step 3. above.
5. The `iree_task_list_pop_front` from step 3. above is already done while a mutex is locked. So I thought that we might solve that by also locking that mutex around the `iree_atomic_task_slist_pop` from step 2. above. Can you suggest another solution? TSan can also be silenced here by making the `->next` dereference in `iree_task_list_pop_front` be relaxed atomic, but I'm concerned that that might allow inconsistency between where both sides read and null that pointer.

Reply:
Ah, it's more complicated still. Some of the races are in `iree_atomic_slist_pop` atomically loading `next`, not in the subsequent nulling. That races with the non-atomic nulling of the `next` on the elements flushed by `iree_atomic_task_slist_flush`. So E. won't fix that. I think the race goes both ways so if we want to fix it by deep-copies, then we need BOTH B. and E., that is, we need ALL functions extracting data from atomic slists to perform deep copies.

Reply:
OK, implemented that idea C. and pushed this commit to this PR, WDYT?
https://github.com/iree-org/iree/pull/9777/commits/5afc4c726bcd8248e38710dc95aae54e45ab9a83

Reply:
I haven't been able to parse through that comment yet, but I think you're still letting TSAN lead you around a bit too much ;)

The update is not correct: the LIFO behavior is intentional and must be preserved.
I'm not sure I understand why you changed that and why you moved it from where it was. Reaching into the queue and touching its internal mutex from another file is not an improvement to me?

Reply:
>What's specific about iree_task_queue_flush_from_lifo_slist is that it performs a shallow-copy of list elements from the atomic slist to a plain list

Flushing does not copy: it is an atomic swap of the entire list. There's no race possible between pop and flush.

> The iree_atomic_task_slist_pop from step 2. above races on the ->next list element pointer with the iree_task_list_pop_front from step 3. above.

Since the slist has been flushed there's no race: anyone else trying to access the slist will see an empty slist and the function that then goes and concatenates the lists is the sole owner.